File: Python/0160_Intersection_2_Linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        def getLength(node):
            length = 0
            while node:
                length += 1
                node = node.next
            return length
        
        lenA, lenB = map(getLength, (headA, headB))
        if lenA > lenB:
            headA, headB = headB, headA
            lenA, lenB = lenB, lenA
        
        diff = lenB - lenA
        nodeA, nodeB = headA, headB
        
        while diff:
            diff -= 1
            nodeB = nodeB.next
        
        while nodeA != nodeB:
            nodeA = nodeA.next
            nodeB = nodeB.next
        return nodeA

File: CPlusPlus/0160_Intersection_2_Linked_list.cpp
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
    auto getLength = [&](ListNode *node) {
        int length = 0;
        while (node != NULL) {
            length += 1;
            node = node->next;
        }
        return length;
    };

    int lenA = getLength(headA), lenB = getLength(headB);
    if (lenA > lenB) {
        swap(headA, headB);
        swap(lenA, lenB);
    }
    int diff = lenB - lenA;
    ListNode *nodeA = headA, *nodeB = headB;

    while (diff--) { // After performing nodeB = nodeB->next, diff -= 1
        nodeB = nodeB->next;
    }

    while (nodeA != nodeB) {
        nodeA = nodeA->next;
        nodeB = nodeB->next;
    }
    return nodeA;
}
